diff -ur watchdog-5.6.orig/src/watchdog.c watchdog-5.6/src/watchdog.c
--- watchdog-5.6.orig/src/watchdog.c	2013-08-06 16:07:28.783256407 +0100
+++ watchdog-5.6/src/watchdog.c	2013-08-06 16:08:16.274256285 +0100
@@ -75,6 +75,11 @@
 #define TESTTIMEOUT	"test-timeout"
 #define HEARTBEAT	"heartbeat-file"
 #define HBSTAMPS	"heartbeat-stamps"
+#define TESTDIR		"test-directory"
+
+#ifndef TESTBIN_PATH
+#define TESTBIN_PATH NULL
+#endif
 
 pid_t pid;
 int softboot = FALSE, watchdog = -1, load = -1, mem = -1, temp = -1;
@@ -217,7 +222,6 @@
     wd_action(keep_alive(), rbinary, NULL, 0);
 }
 
-#ifdef TESTBIN_PATH
 static void do_check2(int res, char *r_specific, char *r_global, char *name)
 {
     wd_action(res, r_specific, name, 1);
@@ -226,7 +230,7 @@
 
 /* Self-repairing binaries list */
 struct list *tr_bin = NULL;
-#endif
+char *test_dir = TESTBIN_PATH;
 
 struct list *file = NULL, *target = NULL, *pidfile = NULL, *iface = NULL;
 char *tbinary, *rbinary, *admin;
@@ -440,6 +444,11 @@
 		else {
 			minpages = atol(line + i);
 		}
+	    } else if (strncmp(line + i, TESTDIR, strlen(TESTDIR)) == 0) {
+		if (spool(line, &i, strlen(TESTDIR)))
+			fprintf(stderr, "Ignoring invalid line in config file: %s ", line);
+		else
+			test_dir = strdup(line + i);
 	    } else {
 		fprintf(stderr, "Ignoring invalid line in config file:\n%s\n", line);
 	    }
@@ -452,7 +461,6 @@
     }
 }
 
-#ifdef TESTBIN_PATH
 static void add_test_binaries(const char *path)
 {
     DIR *d;
@@ -463,6 +471,8 @@
     char fname[PATH_MAX];
     char *fdup;
 
+    if (!path)
+	return;
     ret = stat(path, &sb);
     if (ret < 0)
 	return;
@@ -502,7 +512,6 @@
 	add_list(&tr_bin, fdup);
     } while (1);
 }
-#endif
 
 static void old_option(int c, char *filename)
 {
@@ -591,9 +600,7 @@
     }
 
     read_config(filename, progname);
-#ifdef TESTBIN_PATH
-    add_test_binaries(TESTBIN_PATH);
-#endif
+    add_test_binaries(test_dir);
 
     if (tint < 0)
 	usage();
@@ -898,7 +905,7 @@
 
     /* main loop: update after <tint> seconds */
     while (_running) {
-	wd_action(keep_alive(), rbinary, NULL);
+	wd_action(keep_alive(), rbinary, NULL, 0);
 
 	/* sync system if we have to */
 	do_check(sync_system(sync_it), rbinary, NULL);
@@ -937,6 +944,7 @@
 #ifdef TESTBIN_PATH
 	/* test/repair binaries in the watchdog.d directory */
 	for (act = tr_bin; act != NULL; act = act->next)
+		/* Use version 1 for testbin-path */
 	    do_check2(check_bin(act->name, timeout, 1), act->name, rbinary, NULL);
 #endif
 
